src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -0,0 +1,361 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bytes::Bytes;
+use risingwave_hummock_sdk::key::{user_key, FullKey};
+use risingwave_hummock_sdk::key_range::KeyRange;
+use risingwave_hummock_sdk::HummockEpoch;
+use risingwave_pb::hummock::{Level, LevelType, SstableInfo};
+
+use super::SearchResult;
+use crate::hummock::level_handler::LevelHandler;
+
+const DEFAULT_MAX_COMPACTION_BYTES: u64 = 2 * 1024 * 1024 * 1024; // 2GB
+const DEFAULT_LEVEL0_MAX_FILE_NUMBER: usize = 16;
+const DEFAULT_LEVEL0_TRIGGER_NUMBER: usize = 4;
+
+#[derive(Debug)]
+pub struct TierCompactionPicker {
+    compact_task_id: u64,
+    max_compaction_bytes: u64,
+    level0_max_file_number: usize,
+    level0_trigger_number: usize,
+}
+
+#[derive(Default)]
+pub struct ConflictKeyRange {
+    key_range: Option<KeyRange>,
+}
+
+impl ConflictKeyRange {
+    fn check_table_conflict(&mut self, level_handler: &LevelHandler, table: &SstableInfo) -> bool {
+        let key_range = KeyRange::from(table.key_range.as_ref().unwrap());
+        if level_handler.is_pending_compact(&table.id) {
+            if let Some(range) = self.key_range.as_mut() {
+                range.full_key_extend(&key_range);
+            } else {
+                self.key_range = Some(key_range);
+            }
+            return true;
+        }
+        if let Some(range) = self.key_range.as_ref() {
+            // This file is conflict with previous files in L0.
+            if key_range.full_key_overlap(range) {
+                return true;
```

Comment:
guess incorrect here. Even if `key_range` is not in compaction now, it still needs to be added into `self.key_range`

Reply:
no.. It's not range of current compaction. It's only the range of the previous pending compaction.